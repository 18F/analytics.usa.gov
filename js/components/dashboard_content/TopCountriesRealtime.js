import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import nestCharts from "../../lib/chart_helpers/nest_charts";
import transformers from "../../lib/chart_helpers/transformers";
import { isPartOfUnitedStates } from "../../lib/territories";

/**
 * Retrieves the realtime top countries report from the passed data URL and
 * creates a visualization for the count of users visiting sites for the current
 * agency from each country.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {number} props.refreshSeconds the number of seconds to wait before
 * refreshing chart data.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopCountriesRealtime({ dataHrefBase, refreshSeconds }) {
  const dataURL = `${dataHrefBase}/top-countries-realtime.json`;
  const countriesRef = useRef(null);
  const usTerritoriesRef = useRef(null);
  const internationalVisitsRef = useRef(null);
  const [countryData, setCountryData] = useState(null);
  const [chartsLoaded, setChartsLoaded] = useState(false);

  useEffect(() => {
    const initRealtimeCountriesChart = async () => {
      if (!countryData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setCountryData(data);
        // Refresh data every interval. useEffect will run and update the chart
        // when the state is changed.
        setInterval(() => {
          DataLoader.loadJSON(dataURL).then((data) => {
            setCountryData(data);
            setChartsLoaded(false);
          });
        }, refreshSeconds * 1000);
      } else {
        // Create countries chart
        let chartBuilder = new ChartBuilder();
        await chartBuilder.buildBarChart(
          countriesRef.current,
          countryData,
          (d) => {
            let totalVisits = 0;
            let USVisits = 0;
            d.data.forEach((c) => {
              totalVisits += parseInt(c.active_visitors, 10);
              if (isPartOfUnitedStates(c.country)) {
                USVisits += parseInt(c.active_visitors, 10);
              }
            });
            const international = totalVisits - USVisits;
            const data = {};
            if (USVisits) {
              data["United States &amp; Territories"] = USVisits;
            }
            if (international) {
              data["International"] = international;
            }
            return transformers.findProportionsOfMetricFromValue(
              transformers.listify(data),
            );
          },
        );

        // Create us and territories breakdown chart
        chartBuilder = new ChartBuilder();
        await chartBuilder.buildBarChartWithLabel(
          usTerritoriesRef.current,
          countryData,
          (d) => {
            let values = transformers.findProportionsOfMetric(d.data, (list) =>
              list.map((x) => x.active_visitors),
            );
            values = values.filter((c) => isPartOfUnitedStates(c.country));
            return values.slice(0, 3);
          },
          "country",
        );

        // Create international breakdown chart
        chartBuilder = new ChartBuilder();
        await chartBuilder.buildBarChartWithLabel(
          internationalVisitsRef.current,
          countryData,
          (d) => {
            let values = transformers.findProportionsOfMetric(d.data, (list) =>
              list.map((x) => x.active_visitors),
            );
            values = values.filter((c) => !isPartOfUnitedStates(c.country));
            return values.slice(0, 15);
          },
          "country",
        );

        await setChartsLoaded(true);
      }

      if (chartsLoaded) {
        // nest the US and territories chart inside the "US"
        // chart once they're both rendered
        await d3
          .select(countriesRef.current)
          .call(
            nestCharts,
            "United States &amp; Territories",
            d3.select(usTerritoriesRef.current),
          );

        // nest the international countries chart inside the "International"
        // chart once they're both rendered
        await d3
          .select(countriesRef.current)
          .call(
            nestCharts,
            "International",
            d3.select(internationalVisitsRef.current),
          );
      }
    };
    initRealtimeCountriesChart().catch(console.error);
  }, [countryData, chartsLoaded]);

  return (
    <div>
      <figure id="chart_us" ref={countriesRef}>
        <div className="data bar-chart"></div>
      </figure>

      <figure
        id="chart_us_and_territories"
        className="hide"
        ref={usTerritoriesRef}
      >
        <h4>United States &amp; Territories</h4>
        <div className="data bar-chart"></div>
      </figure>

      <figure
        id="chart_countries"
        className="hide"
        ref={internationalVisitsRef}
      >
        <h4>International</h4>
        <div className="data bar-chart"></div>
      </figure>
    </div>
  );
}

TopCountriesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  refreshSeconds: PropTypes.number.isRequired,
};

export default TopCountriesRealtime;
