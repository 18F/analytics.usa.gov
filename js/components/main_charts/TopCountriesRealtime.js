import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import nestCharts from "../../lib/chart_helpers/nest_charts";
import renderBlock from "../../lib/chart_helpers/renderblock";
import transformers from "../../lib/chart_helpers/transformers";
import { isPartOfUnitedStates } from "../../lib/territories";

/**
 * Retrieves the realtime top countries report from the passed data URL and
 * creates a visualization for the count of users visiting sites for the current
 * agency from each country.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 */
function TopCountriesRealtime({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-countries-realtime.json`;
  const countriesRef = useRef(null);
  const usTerritoriesRef = useRef(null);
  const internationalVisitsRef = useRef(null);

  useEffect(() => {
    const initRealtimeCountriesChart = async () => {
      // Create countries chart
      await d3
        .select(countriesRef.current)
        .datum({
          source: dataURL,
          block: countriesRef.current,
        })
        .call(
          renderBlock.buildBarChart((d) => {
            let totalVisits = 0;
            let USVisits = 0;
            d.data.forEach((c) => {
              totalVisits += parseInt(c.active_visitors, 10);
              if (isPartOfUnitedStates(c.country)) {
                USVisits += parseInt(c.active_visitors, 10);
              }
            });
            const international = totalVisits - USVisits;
            const data = {
              "United States &amp; Territories": USVisits,
              International: international,
            };
            return transformers.findProportionsOfMetricFromValue(
              transformers.listify(data),
            );
          }, "country"),
        );

      // Create us and territories breakdown chart
      await d3
        .select(usTerritoriesRef.current)
        .datum({
          source: dataURL,
          block: usTerritoriesRef.current,
        })
        .call(
          renderBlock.buildBarChartWithLabel((d) => {
            let values = transformers.findProportionsOfMetric(d.data, (list) =>
              list.map((x) => x.active_visitors),
            );
            values = values.filter((c) => isPartOfUnitedStates(c.country));
            return values.slice(0, 3);
          }, "country"),
        );

      // Create international breakdown chart
      await d3
        .select(internationalVisitsRef.current)
        .datum({
          source: dataURL,
          block: internationalVisitsRef.current,
        })
        .call(
          renderBlock.buildBarChartWithLabel((d) => {
            let values = transformers.findProportionsOfMetric(d.data, (list) =>
              list.map((x) => x.active_visitors),
            );
            values = values.filter((c) => !isPartOfUnitedStates(c.country));
            return values.slice(0, 15);
          }, "country"),
        );

      // Sleep for half a second because the above charts are still not loaded
      // for some reason.  TODO: figure out why
      const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
      await delay(500);

      // nest the US and territories chart inside the "US"
      // chart once they're both rendered
      await d3
        .select(countriesRef.current)
        .call(
          nestCharts,
          "United States &amp; Territories",
          d3.select(usTerritoriesRef.current),
        );

      // nest the international countries chart inside the "International"
      // chart once they're both rendered
      await d3
        .select(countriesRef.current)
        .call(
          nestCharts,
          "International",
          d3.select(internationalVisitsRef.current),
        );

      return;
    };
    initRealtimeCountriesChart().catch(console.error);
  }, []);

  return (
    <>
      <figure id="chart_us" ref={countriesRef}>
        <div className="data bar-chart"></div>
      </figure>

      <figure
        id="chart_us_and_territories"
        className="hide"
        ref={usTerritoriesRef}
      >
        <h4>United States &amp; Territories</h4>
        <div className="data bar-chart"></div>
      </figure>

      <figure
        id="chart_countries"
        className="hide"
        ref={internationalVisitsRef}
      >
        <h4>International</h4>
        <div className="data bar-chart"></div>
      </figure>
    </>
  );
}

TopCountriesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopCountriesRealtime;
