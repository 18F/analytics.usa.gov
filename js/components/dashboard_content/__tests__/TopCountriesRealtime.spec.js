import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopCountriesRealtime from "../TopCountriesRealtime";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("TopCountriesRealtime", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopCountriesRealtime
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "top-countries-realtime",
        query: {},
        meta: {
          name: "Top Countries",
          description: "Top countries for active onsite users.",
        },
        data: [
          {
            country: "United States",
            active_visitors: "803775",
          },
          {
            country: "India",
            active_visitors: "11231",
          },
          {
            country: "Canada",
            active_visitors: "10002",
          },
          {
            country: "United Kingdom",
            active_visitors: "8097",
          },
          {
            country: "Mexico",
            active_visitors: "7900",
          },
          {
            country: "Colombia",
            active_visitors: "3966",
          },
          {
            country: "Philippines",
            active_visitors: "3831",
          },
          {
            country: "Germany",
            active_visitors: "3163",
          },
          {
            country: "Puerto Rico",
            active_visitors: "2785",
          },
          {
            country: "France",
            active_visitors: "2453",
          },
          {
            country: "Spain",
            active_visitors: "2230",
          },
          {
            country: "Argentina",
            active_visitors: "1912",
          },
          {
            country: "Brazil",
            active_visitors: "1879",
          },
          {
            country: "Peru",
            active_visitors: "1579",
          },
          {
            country: "Pakistan",
            active_visitors: "1543",
          },
          {
            country: "Italy",
            active_visitors: "1527",
          },
          {
            country: "Chile",
            active_visitors: "1497",
          },
          {
            country: "Indonesia",
            active_visitors: "1392",
          },
        ],
        totals: {},
        taken_at: "2024-01-05T16:05:49.789Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopCountriesRealtime
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
      await waitFor(() => screen.getByText("Spain"));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", async () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopCountriesRealtime
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
