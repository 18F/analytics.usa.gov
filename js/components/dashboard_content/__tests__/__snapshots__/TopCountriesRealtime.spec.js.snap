// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopCountriesRealtime when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <div>
    <figure
      id="chart_us"
    >
      <div
        class="data bar-chart"
      >
        <div
          class="bin"
        >
          <div
            class="label"
          >
            United States & Territories
          </div>
          <div
            class="value"
          >
            92.8%
          </div>
          <div
            class="bar"
            style="width: 0.3712000000000001%;"
          />
          <figure
            class=""
            id="chart_us_and_territories"
          >
            <h4>
              United States & Territories
            </h4>
            <div
              class="data bar-chart"
            >
              <div
                class="bin"
                data-share="92.46583615851056"
              >
                <div
                  class="label"
                >
                  United States
                </div>
                <div
                  class="value"
                >
                  92.5%
                </div>
                <div
                  class="bar"
                  style="width: 92.5%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.32038487599322185"
              >
                <div
                  class="label"
                >
                  Puerto Rico
                </div>
                <div
                  class="value"
                >
                  0.3%
                </div>
                <div
                  class="bar"
                  style="width: 0.3%;"
                />
              </div>
            </div>
          </figure>
        </div>
        <div
          class="bin"
        >
          <div
            class="label"
          >
            International
          </div>
          <div
            class="value"
          >
            7.2%
          </div>
          <div
            class="bar"
            style="width: 0.028800000000000006%;"
          />
          <figure
            class=""
            id="chart_countries"
          >
            <h4>
              International
            </h4>
            <div
              class="data bar-chart"
            >
              <div
                class="bin"
                data-share="1.2920080941758978"
              >
                <div
                  class="label"
                >
                  India
                </div>
                <div
                  class="value"
                >
                  1.3%
                </div>
                <div
                  class="bar"
                  style="width: 1.3%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.978640624802276"
              >
                <div
                  class="label"
                >
                  Canada
                </div>
                <div
                  class="value"
                >
                  1%
                </div>
                <div
                  class="bar"
                  style="width: 1.0%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.9314744491623401"
              >
                <div
                  class="label"
                >
                  United Kingdom
                </div>
                <div
                  class="value"
                >
                  0.9%
                </div>
                <div
                  class="bar"
                  style="width: 0.9%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.9088116769646151"
              >
                <div
                  class="label"
                >
                  Mexico
                </div>
                <div
                  class="value"
                >
                  0.9%
                </div>
                <div
                  class="bar"
                  style="width: 0.9%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.4562464697267928"
              >
                <div
                  class="label"
                >
                  Colombia
                </div>
                <div
                  class="value"
                >
                  0.5%
                </div>
                <div
                  class="bar"
                  style="width: 0.5%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.4407161436014481"
              >
                <div
                  class="label"
                >
                  Philippines
                </div>
                <div
                  class="value"
                >
                  0.4%
                </div>
                <div
                  class="bar"
                  style="width: 0.4%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.363869789144187"
              >
                <div
                  class="label"
                >
                  Germany
                </div>
                <div
                  class="value"
                >
                  0.4%
                </div>
                <div
                  class="bar"
                  style="width: 0.4%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.28219177767015197"
              >
                <div
                  class="label"
                >
                  France
                </div>
                <div
                  class="value"
                >
                  0.3%
                </div>
                <div
                  class="bar"
                  style="width: 0.3%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.25653797970013814"
              >
                <div
                  class="label"
                >
                  Spain
                </div>
                <div
                  class="value"
                >
                  0.3%
                </div>
                <div
                  class="bar"
                  style="width: 0.3%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.2199554337159929"
              >
                <div
                  class="label"
                >
                  Argentina
                </div>
                <div
                  class="value"
                >
                  0.2%
                </div>
                <div
                  class="bar"
                  style="width: 0.2%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.21615913177424195"
              >
                <div
                  class="label"
                >
                  Brazil
                </div>
                <div
                  class="value"
                >
                  0.2%
                </div>
                <div
                  class="bar"
                  style="width: 0.2%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.18164729594014267"
              >
                <div
                  class="label"
                >
                  Peru
                </div>
                <div
                  class="value"
                >
                  0.2%
                </div>
                <div
                  class="bar"
                  style="width: 0.2%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.17750587564005074"
              >
                <div
                  class="label"
                >
                  Pakistan
                </div>
                <div
                  class="value"
                >
                  0.2%
                </div>
                <div
                  class="bar"
                  style="width: 0.2%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.17566524439556547"
              >
                <div
                  class="label"
                >
                  Italy
                </div>
                <div
                  class="value"
                >
                  0.2%
                </div>
                <div
                  class="bar"
                  style="width: 0.2%;"
                />
              </div>
              <div
                class="bin"
                data-share="0.17221406081215554"
              >
                <div
                  class="label"
                >
                  Chile
                </div>
                <div
                  class="value"
                >
                  0.2%
                </div>
                <div
                  class="bar"
                  style="width: 0.2%;"
                />
              </div>
            </div>
          </figure>
        </div>
      </div>
    </figure>
  </div>
</DocumentFragment>
`;

exports[`TopCountriesRealtime when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div>
    <figure
      id="chart_us"
    >
      <div
        class="data bar-chart"
      />
    </figure>
    <figure
      class="hide"
      id="chart_us_and_territories"
    >
      <h4>
        United States & Territories
      </h4>
      <div
        class="data bar-chart"
      />
    </figure>
    <figure
      class="hide"
      id="chart_countries"
    >
      <h4>
        International
      </h4>
      <div
        class="data bar-chart"
      />
    </figure>
  </div>
</DocumentFragment>
`;

exports[`TopCountriesRealtime when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div>
    <figure
      id="chart_us"
    >
      <div
        class="data bar-chart"
      />
    </figure>
    <figure
      class="hide"
      id="chart_us_and_territories"
    >
      <h4>
        United States & Territories
      </h4>
      <div
        class="data bar-chart"
      />
    </figure>
    <figure
      class="hide"
      id="chart_countries"
    >
      <h4>
        International
      </h4>
      <div
        class="data bar-chart"
      />
    </figure>
  </div>
</DocumentFragment>
`;
